docs/examples/move/nft_marketplace/sources/marketplace_extension.move

```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// Setup item royalty percentage
+    /// - amount_bp - the percentage of the purchase price to be paid as a
+    /// fee, denominated in basis points (100_00 = 100%, 1 = 0.01%).
+    /// - min_amount - the minimum amount to be paid as a fee if the relative
+    /// amount is lower than this setting.
+    public fun setup_royalties<T: key + store>(policy: &mut TransferPolicy<T>, cap: &TransferPolicyCap<T>, amount_bp: u16, min_amount: u64) {
+        royalty_rule::add<T>(policy, cap, amount_bp, min_amount);
+    }
+
+    /// Buy listed item with the indicated price and pay royalties if needed
+    public fun buy_item<T: key + store>(kiosk: &mut Kiosk, policy: &mut TransferPolicy<T>, item_id: object::ID, mut payment: Coin<IOTA>, ctx: &mut TxContext): T {
+        assert!(kiosk_extension::is_installed<Marketplace>(kiosk), EExtensionNotInstalled);
+
+        // Get item price
+        let ItemPrice { price } = take_from_bag<T, Listed>(kiosk,  Listed { id: item_id });
+
+        // Compute the value of the coin in input
+        let payment_amount_value = payment.value();
+
+        // If the payment_amount_value is less than the item price, the request is invalid.
+        assert!(payment_amount_value >= price, ENotEnoughPaymentAmount);
+
+        // Prepare the payment coin for the purchase (if no royalties are present then the
+        // remaining balance will be 0 after this operation)
+        let coin_price = payment.split(price, ctx);
+
+        // Purchase and create the transfer request
+        let (item, mut transfer_request) = purchase(kiosk, item_id, coin_price);
+
+        // If the royalty is present, then update the request with a royalty payment
+        if (policy.has_rule<T, RoyaltyRule>()) { 
+            let royalties_value = royalty_rule::fee_amount(policy, price);
+            assert!(payment_amount_value == price + royalties_value, EWrongPaymentRoyalties);
+            royalty_rule::pay(policy, &mut transfer_request, payment);
+        } else {
+            // Else clean the input coin (if the input payment amount is not exact, this will fail)
+            payment.destroy_zero();
+        };
+
+        // Confirm the request
+        transfer_policy::confirm_request(policy, transfer_request);
+        item
+    }
+
+
+    public fun set_price<T: key + store>(
+        kiosk: &mut Kiosk,
+        cap: &KioskOwnerCap,
+        item: T,
+        price: u64) {
+        assert!(kiosk_extension::is_installed<Marketplace>(kiosk), EExtensionNotInstalled);
+
+        let id = object::id(&item);
+        kiosk.place_and_list<T>(cap, item, price);
+
+        let item_price = ItemPrice {
+            price,
+        };
+
+        place_in_bag<T, Listed>(kiosk, Listed { id }, item_price);
+    }
+
+
+    public fun get_item_price<T: key + store>(
+        kiosk: &Kiosk,
+        item_id: ID,
+    ) : u64 {
+        let storage_ref = kiosk_extension::storage(Marketplace {}, kiosk);
```

Comment:
Just wondering, why this function receives an extension instance?
Kind of to be sure that it is called from the extension module?

Reply:
Internally it needs only the type but not the instance. So, it is interesting why we need the `_ext: Ext` param.

Reply:
We have to have publisher anyway, publisher is required for TransferPolicy creation
So semantically and logically publisher should be created inside clothing_store as we use these items for marketplace stuff.

Reply:
> Kind of to be sure that it is called from the extension module? 
I guess it’s an answer, anyway it’s inner implementation of kiosk functionality.
Maybe it’s kind of obsolete behavior, but for now we have to follow it